packages/dev-server/src/middleware/JSInspectorMiddleware.ts

```diff
@@ -0,0 +1,127 @@
+import type { NextHandleFunction } from 'connect';
+import fs from 'fs-extra';
+import type { IncomingMessage, ServerResponse } from 'http';
+import fetch from 'node-fetch';
+import open from 'open';
+import path from 'path';
+import { TLSSocket } from 'tls';
+import { URL } from 'url';
+
+export default function JSInspectorMiddleware(): NextHandleFunction {
+  return async function (req: IncomingMessage, res: ServerResponse, next: (err?: Error) => void) {
+    const { origin, searchParams } = new URL(req.url ?? '/', getRequestBase(req));
+    const applicationId = searchParams.get('applicationId');
+    if (!applicationId) {
+      res.writeHead(400).end('Missing applicationId');
+      return;
+    }
+
+    const target = await queryInspectorTargetAsync(origin, applicationId);
+    if (!target) {
+      res.writeHead(404).end('Unable to find inspector target from metro-inspector-proxy');
+      return;
+    }
+
+    if (req.method === 'GET') {
+      const data = JSON.stringify(target);
+      res.writeHead(200, {
+        'Content-Type': 'application/json; charset=UTF-8',
+        'Cache-Control': 'no-cache',
+        'Content-Length': data.length.toString(),
+      });
+      res.end(data);
+    } else if (req.method === 'POST' || req.method === 'PUT') {
+      // To update devtoolsFrontendRev, find the full commit hash in the url:
+      // https://chromium.googlesource.com/chromium/src.git/+log/refs/tags/{CHROME_VERSION}/chrome/VERSION
+      //
+      // 1. Replace {CHROME_VERSION} with the target chrome version
+      // 2. Click the first log item in the webpage
+      // 3. The full commit hash is the desired revision
+      const devtoolsFrontendRev = 'e3cd97fc771b893b7fd1879196d1215b622c2bed'; // Chrome 90.0.4430.212
+
+      const urlBase = `https://chrome-devtools-frontend.appspot.com/serve_rev/@${devtoolsFrontendRev}/inspector.html`;
+      const ws = target.webSocketDebuggerUrl.replace('ws://[::]:', 'localhost:');
+      const url = `${urlBase}?experiments=true&v8only=true&ws=${encodeURIComponent(ws)}`;
+      launchChromiumAsync(url);
+      res.end();
+    } else {
+      res.writeHead(405);
+    }
+  };
+}
+
+interface MetroInspectorProxyApp {
+  description: string;
+  devtoolsFrontendUrl: string;
+  faviconUrl: string;
+  id: string;
+  title: string;
+  type: 'node';
+  vm: 'Hermes' | "don't use";
+  webSocketDebuggerUrl: string;
+}
+
+async function queryInspectorTargetAsync(
+  origin: string,
+  appId: string
+): Promise<MetroInspectorProxyApp | null> {
+  const resp = await fetch(`${origin}/json/list`);
+  const apps: MetroInspectorProxyApp[] = await resp.json();
+
+  let target: MetroInspectorProxyApp | null = null;
+  for (const app of apps) {
+    if (app.description === appId && app.vm !== "don't use") {
```

Comment:
haha, why "don't use"?

Reply:
This is from [metro-inspector-proxy](https://github.com/facebook/metro/blob/3aca11697323dacef70bb914399e5e064fc74870/packages/metro-inspector-proxy/src/Device.js#L135-L143).
seems to support inspecting for live reload or fast refresh, but still in experiment.
i am not pretty sure how stable it is, so to exclude it in the mean time.